### ISSUE TITLE
"is_training" parameter in the batch normalization layers

### ISSUE THREAD

**Issue by gabrielazg** [arahusky/Tensorflow-Segmentation#15]
Hello! I might be missing something about the way the batch normalization layers are defined and therefore I would like to pose a question, particularly regarding the parameter "is_training" in tf.contrib.layers.batch_norm. I noticed that the parameter is not set when the batch normalization layers are defined, which makes it take the default value is_training=True always. However I believe it should be set as **is_training=False when we are doing inference**, after the training process is concluded. Does it make sense? If so, then the existing is_training placeholder can be linked to the batch_norm layers, which would make it simple to set it True or False for training operations and inference operations, respectively. Let me know what you think. Many thanks!

As additional reference, the Tensorflow documentation defines the "is_training" parameter of tf.contrib.layers.batch_norm as "whether or not the layer is in training mode. In training mode it would accumulate the statistics of the moments into moving_mean and moving_variance using an exponential moving average with the given decay. When it is not in training mode then it would use the values of the moving_mean and the moving_variance".

**Comment by arahusky** [comment]
Hello @gabrielazg ,
thanks for your valuable comment. You're absolutely right. The implementation should set is_training parameter to False when inferring. Moreover, either updates_collections of batch_norm method should be set None or control_dependendency must be set in training (see http://ruishu.io/2016/12/27/batchnorm/). 

It is not clear to me, why does inferring work even if is_training is set to True (maybe too small batch?). I will try to update code soon and see the difference.

**Comment by gabrielazg** [comment]
Exactly, I agree with your further comments. This is a good reference: http://ruishu.io/2016/12/27/batchnorm/. Regardless of the boolean in is_training, the batch normalisation layer will always do the batch normalisation, so it is up to us to control if it should also "update population statistics approximation via moving average of mini-batch statistics" (learning mode, i.e. is_training=True) or not (inference mode, i.e. is_training=False).

**Comment by qwangku** [comment]
Will excluding "moving_mean" and "moving_variance" from var_to_update have the same effect as is_training==False?

**Comment by davidbernat** [comment]
Can you speak to this user's issue? 

This seems to indicate a bug/unclear use case for setting is_training to False during inference. Doing so creates a situation in which all predictions are the same value. I am running into this problem currently. 

https://stackoverflow.com/questions/46153978/tensorflow-slim-model-trains-but-always-predicts-the-same-when-evaluating

**Comment by ShiyuZhangUST** [comment]
> Can you speak to this user's issue?
> 
> This seems to indicate a bug/unclear use case for setting is_training to False during inference. Doing so creates a situation in which all predictions are the same value. I am running into this problem currently.
> 
> https://stackoverflow.com/questions/46153978/tensorflow-slim-model-trains-but-always-predicts-the-same-when-evaluating

I got this problem today and solved it, answered on the stackoverflow link.
